pkg/client/query/cache/memory.go

```diff
@@ -0,0 +1,304 @@
+	if c.config.historical {
+		return c.SetAtHeight(key, value, c.latestHeight.Load())
+	}
+
+	if c.config.MaxKeys > 0 && int64(len(c.items)) >= c.config.MaxKeys {
+		c.evict()
+	}
+
+	c.itemsMu.Lock()
+	defer c.itemsMu.Unlock()
+
+	c.items[key] = cacheItem[T]{
+		value:     value,
+		timestamp: time.Now(),
+	}
+
+	return nil
+}
+
+// SetAtHeight adds or updates the historical value in the cache for the given key,
+// and at the given height. If the cache is not configured for historical mode, it
+// returns an error.
+func (c *InMemoryCache[T]) SetAtHeight(key string, value T, setHeight int64) error {
+	if !c.config.historical {
+		return ErrHistoricalModeNotEnabled
+	}
+
+	// Update c.latestHeight if the given setHeight is newer (higher).
+	latestHeight := c.latestHeight.Load()
+	if setHeight > latestHeight {
+		// NB: Only update if c.latestHeight hasn't changed since we loaded it above.
+		c.latestHeight.CompareAndSwap(latestHeight, setHeight)
+	}
+
+	c.itemsMu.Lock()
+	defer c.itemsMu.Unlock()
+
+	var itemHistory cacheItemHistory[T]
+	if itemHistoryAny, exists := c.items[key]; exists {
+		itemHistory = itemHistoryAny.(cacheItemHistory[T])
+	} else {
+		itemsByHeight := make(map[int64]cacheItem[T])
+		itemHistory = cacheItemHistory[T]{
+			sortedDescHeights: make([]int64, 0),
+			itemsByHeight:     itemsByHeight,
+		}
+	}
+
+	// Update sortedDescHeights and ensure the list is sorted in descending order.
+	if _, setHeightExists := itemHistory.itemsByHeight[setHeight]; !setHeightExists {
+		itemHistory.sortedDescHeights = append(itemHistory.sortedDescHeights, setHeight)
+		sort.Slice(itemHistory.sortedDescHeights, func(i, j int) bool {
+			return itemHistory.sortedDescHeights[i] > itemHistory.sortedDescHeights[j]
+		})
+	}
+
+	// Prune historical values for this key, where the setHeight
+	// is oder than the configured pruneOlderThan.
+	if c.config.pruneOlderThan > 0 {
+		lenCachedHeights := int64(len(itemHistory.sortedDescHeights))
+		for heightIdx := lenCachedHeights - 1; heightIdx >= 0; heightIdx-- {
+			cachedHeight := itemHistory.sortedDescHeights[heightIdx]
+
+			// DEV_NOTE: Since the list is sorted, and we're iterating from lowest
+			// (oldest) to highest (youngest) height, once we encounter a cachedHeight
+			// that is younger than the configured pruneOlderThan, ALL subsequent
+			// heights SHOULD also be younger than the configured pruneOlderThan.
+			if setHeight-cachedHeight <= c.config.pruneOlderThan {
+				itemHistory.sortedDescHeights = itemHistory.sortedDescHeights[:heightIdx+1]
+				break
+			}
+
+			delete(itemHistory.itemsByHeight, cachedHeight)
+		}
+	}
+
+	itemHistory.itemsByHeight[setHeight] = cacheItem[T]{
+		value:     value,
+		timestamp: time.Now(),
+	}
+
+	c.items[key] = itemHistory
+
+	return nil
+}
+
+// Delete removes an item from the cache.
+func (c *InMemoryCache[T]) Delete(key string) {
+	c.itemsMu.Lock()
+	defer c.itemsMu.Unlock()
+
+	delete(c.items, key)
```

Comment:
What if we delete the (key,value) pair for `latestHeight`?

Reply:
Could you elaborate, I'm not quite following.

Reply:
Yea, this comment doesn't make sense if you don't know where I was looking or what I was thinking...

If we delete the key for the latest height, do we need to also make an appropriate call to `c.latestHeight.Store(??)`?

If this is indeed the case, please add a unit test for it.

If not, would just appreciate a quick (no need for anything big) explanation.

Reply:
Gotcha, good question. I don't think we need to update `c.latestHeight`, regardless of which key is being deleted.  For `Get`s, this should be completely transparent to the caller; for `Set`s, it means that if you call `Delete`, followed by `Set` (as opposed to `SetAtHeight`), then the height you're setting will be the last `c.latestHeight`. I think this behavior is correct. 

> If not, would just appreciate a quick (no need for anything big) explanation.
I did see :point_up_2: but had to think it through for myself anyways:

```mermaid
flowchart LR

subgraph hc2[Historical Cache]
    lh2["latestHeight: 2"]
    subgraph k2_2h["key2 history"]
        k2_2h1["height: 1; value 100"]
    end
    subgraph k2_1h["key1 history"]
        k2_1h1["height: 1; value 10"]
        k2_1h2["height: 2; value 20"]
    end
end

subgraph hc3[Historical Cache]
    lh3["latestHeight: 2"]
    subgraph k3_2h["key2 history"]
        k3_2h1["height: 1; value 100"]
    end
end


subgraph hc4[Historical Cache]
    lh4["latestHeight: 2"]
    subgraph k4_1h["key1 history"]
        k4_1h2["height: 2; value 30"]
    end
end

hc2 --"Delete(key1)"--> hc3
hc3 --"Set(key1, 30)"--> hc4

lh2 --> k2_1h2
lh4 --> k4_1h2
```

Reply:
Given the explanation and diagram, I don't fully understand the need for `latestHeight` at all. In fact, I think we should delete it now if the behaviour you described is expected.

Reply:
I neglected to add a comment to this thread but after discussing with @red-0ne, we decided that it should be an error to call `Set()` on a `HistoricalQueryCache`. I will update the interface and `inMemoryCache#Set()` to reflect this. In this context, `latestVersion` is no longer necessary and can be removed.

Reply:
```
// QueryCache is a key/value store style interface for a cache of a single type.
// It is intended to be used to cache query responses (or derivatives thereof),
// where each key uniquely indexes the most recent query response.
type QueryCache[T any] interface {
	Get(key string) (T, error)
	Set(key string, value T) error
	Delete(key string)
	Clear()
}

// HistoricalQueryCache extends QueryCache to support getting and setting values
// at multiple heights for a given key.
type HistoricalQueryCache[T any] interface {
	// GetLatestVersion retrieves the historical value with the highest version number.
	GetLatestVersion(key string) (T, error)
	// GetAsOfVersion retrieves the nearest value <= the specified version number.
	GetAsOfVersion(key string, version int64) (T, error)
	// SetAsOfVersion adds or updates a value at a specific version number.
	SetAsOfVersion(key string, value T, version int64) error
}
```
(class diagrams also updated)